CAPH Controller panics during HetznerCluster reconciliation

**Description:**

When attempting to provision a Kubernetes cluster on Hetzner Cloud using Cluster API (CAPI) with Cluster API Provider Hetzner (CAPH), the CAPH controller (`caph-controller-manager`) enters a CrashLoopBackOff state shortly after the `HetznerCluster` resource is applied.

The controller logs show a persistent `panic: runtime error: invalid memory address or nil pointer dereference` occurring specifically within the load balancer related code (`github.com/syself/cluster-api-provider-hetzner/pkg/services/hcloud/loadbalancer.createOptsFromSpec`).

This panic prevents the controller from reconciling the `HetznerCluster` object and subsequently provisioning any infrastructure resources on Hetzner Cloud. The CAPH webhook service also becomes unavailable as a result, leading to webhook errors when applying/updating resources.

**Steps to reproduce:**

1.  Set up a local Kubernetes cluster (using Kind) to serve as the Cluster API management cluster.
2.  Install Cluster API core components.
3.  Install Cluster API Bootstrap Provider Talos (CABPT) and Cluster API Control Plane Provider Talos (CACPPT).
4.  Install Cluster API Provider Hetzner (CAPH) version [Insert CAPH Version Here, e.g., from Image Tag].
5.  Create a Kubernetes Secret in the management cluster with the Hetzner Cloud API token.
6.  Apply a manifest defining a CAPI Cluster and associated Hetzner/Talos infrastructure resources (see Configuration section below). This manifest includes a `HetznerCluster` resource with `controlPlaneLoadBalancer.enabled: false` and a placeholder `controlPlaneEndpoint`.
7.  Observe the logs of the `caph-controller-manager` pod in the `caph-system` namespace.

**Expected behavior:**

The CAPH controller should successfully reconcile the `HetznerCluster` resource, interact with the Hetzner Cloud API to provision the necessary infrastructure (like the private network and placement groups), and proceed without panicking.

**Actual behavior:**

The `caph-controller-manager` pod crashes repeatedly and enters `CrashLoopBackOff`. Logs show a `panic: runtime error: invalid memory address or nil pointer dereference` traceback related to load balancer logic. The `HetznerCluster` resource remains in a state indicating provisioning issues, and no infrastructure is provisioned on Hetzner Cloud.

**Environment:**

.env
```sh
export CLUSTER_NAME="my-cluster" # Choose a unique name for your cluster
export HCLOUD_REGION="fsn1" # Your desired server region (e.g., fsn1)
export HCLOUD_NETWORK_ZONE="eu-central" # Network Zone (must be one of the supported values from the error)
export HCLOUD_SSH_KEY="cluster" # The name of the SSH key in your Hetzner project
export HCLOUD_CONTROL_PLANE_MACHINE_TYPE="cax11" # Server type for control planes
export HCLOUD_WORKER_MACHINE_TYPE="cax21" # Server type for workers
export HCLOUD_TALOS_IMAGE_ID="ce4c980550dd2ab1b17bbf2b08801c7eb59418eafe8f279833297925d67c7515" # The Image ID you found in Step 2a
export CONTROL_PLANE_MACHINE_COUNT=1 # Number of control plane nodes (recommended 3 for HA)
export WORKER_MACHINE_COUNT=1 # Initial number of worker nodes (can be 0 for autoscaling later)
export KUBERNETES_VERSION="v1.31.8" # The Kubernetes version
export TALOS_VERSION="v1.10" # The Talos version compatible with your K8s version
```

hetzner-talos-cluster.yaml 
```yaml
# gitops/infrastructure/hetzner-talos-cluster.yaml

---
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: ${CLUSTER_NAME}
  namespace: default # Or your target namespace
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
        - 10.244.0.0/16 # Default for many CNIs, adjust if needed
    services:
      cidrBlocks:
        - 10.96.0.0/12 # Default for many CNIs, adjust if needed
  controlPlaneRef:
    apiVersion: controlplane.cluster.x-k8s.io/v1alpha3
    kind: TalosControlPlane
    name: ${CLUSTER_NAME}-controlplane
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: HetznerCluster
    name: ${CLUSTER_NAME}
---
# HetznerCluster resource - using v1beta1
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: HetznerCluster
metadata:
  name: ${CLUSTER_NAME}
  namespace: default
spec:
  hcloudNetwork:
    enabled: true
    networkZone: ${HCLOUD_NETWORK_ZONE} # e.g., "eu-central"
    # cidrBlock: 10.0.0.0/16 # Optional: specify if you want a specific private network CIDR
    # subnetCidrBlock: 10.0.0.0/24 # Optional: specify if you want a specific subnet CIDR

  controlPlaneLoadBalancer:
    enabled: true
    region: ${HCLOUD_REGION} # not in docs but needed
    type: lb11 # Specify the Load Balancer type (e.g., lb11, lb21)

  sshKeys:
    hcloud:
      - name: ${HCLOUD_SSH_KEY}

  hetznerSecretRef: # Reference the secret containing Hetzner token
    name: hetzner # Name of the secret in the management cluster (default namespace)
    key:
      hcloudToken: hcloud # Assuming your secret key is named 'hcloud'

  hcloudPlacementGroups:
  - name: ${CLUSTER_NAME}-controlplane
    type: spread
  - name: ${CLUSTER_NAME}-worker
    type: spread

  controlPlaneRegions:
    - "${HCLOUD_REGION}" # Use your defined region here, must be an array

---
# Template for Control Plane Machines' infrastructure (HCloud specific) - using v1beta1
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: HCloudMachineTemplate
metadata:
  name: ${CLUSTER_NAME}-controlplane
  namespace: default
spec:
  template:
    spec:
      type: ${HCLOUD_CONTROL_PLANE_MACHINE_TYPE}
      imageName: ${HCLOUD_TALOS_IMAGE_ID} # Use the Image/Snapshot ID
      placementGroupName: ${CLUSTER_NAME}-controlplane
---
# Template for Worker Machines' infrastructure (HCloud specific) - using v1beta1
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: HCloudMachineTemplate
metadata:
  name: ${CLUSTER_NAME}-worker
  namespace: default
spec:
  template:
    spec:
      type: ${HCLOUD_WORKER_MACHINE_TYPE}
      imageName: ${HCLOUD_TALOS_IMAGE_ID} # Use the Image/Snapshot ID
      placementGroupName: ${CLUSTER_NAME}-worker
---
# Talos bootstrap and control plane configuration for Control Plane nodes
# Use v1alpha3 for TalosControlPlane
apiVersion: controlplane.cluster.x-k8s.io/v1alpha3
kind: TalosControlPlane
metadata:
  name: ${CLUSTER_NAME}-controlplane
  namespace: default
spec:
  replicas: ${CONTROL_PLANE_MACHINE_COUNT}
  version: ${KUBERNETES_VERSION}
  infrastructureTemplate:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: HCloudMachineTemplate
    name: ${CLUSTER_NAME}-controlplane
  controlPlaneConfig:
    controlplane:
      generateType: controlplane
      talosVersion: ${TALOS_VERSION}
      strategicPatches:
        - |
          cluster:
            externalCloudProvider:
              enabled: true
        - |
          cluster:
            network:
              cni: null
---
# MachineDeployment for Worker nodes - using v1beta1
apiVersion: cluster.x-k8s.io/v1beta1
kind: MachineDeployment
metadata:
  name: ${CLUSTER_NAME}-worker-pool
  namespace: default
spec:
  clusterName: ${CLUSTER_NAME}
  replicas: ${WORKER_MACHINE_COUNT}
  selector:
    matchLabels: null
  template:
    spec:
      bootstrap:
        configRef:
          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
          kind: TalosConfigTemplate
          name: ${CLUSTER_NAME}-worker
      clusterName: ${CLUSTER_NAME}
      version: ${KUBERNETES_VERSION}
      infrastructureRef:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
        kind: HCloudMachineTemplate
        name: ${CLUSTER_NAME}-worker
---
# Talos bootstrap configuration template for Worker nodes
# Use v1alpha3 for TalosConfigTemplate
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
kind: TalosConfigTemplate
metadata:
  name: ${CLUSTER_NAME}-worker
  namespace: default
spec:
  template:
    spec:
      # These fields were correct as per the TalosConfigTemplate schema
      generateType: worker
      talosVersion: ${TALOS_VERSION}
      strategicPatches:
        - |
          cluster:
            externalCloudProvider:
              enabled: true
        - |
          cluster:
            network:
              cni: null
---

```


**Logs:**

```text
{"level":"INFO","time":"2025-05-07T14:29:15.090Z","file":"controllers/hcloudmachinetemplate_controller.go:92","message":"HCloudMachineTemplate is missing ownerRef to cluster or cluster does not exist default/my-cluster-controlplane","controller":"hcloudmachinetemplate","controllerGroup":"infrastructure.cluster.x-k8s.io","controllerKind":"HCloudMachineTemplate","HCloudMachineTemplate":{"name":"my-cluster-controlplane","namespace":"default"},"namespace":"default","name":"my-cluster-controlplane","reconcileID":"67aa9fa6-e7de-4148-8a32-2240d1ca9ab8","HCloudMachineTemplate":{"name":"my-cluster-controlplane","namespace":"default"}}
{"level":"INFO","time":"2025-05-07T14:29:15.172Z","file":"controllers/hcloudmachinetemplate_controller.go:92","message":"HCloudMachineTemplate is missing ownerRef to cluster or cluster does not exist default/my-cluster-controlplane","controller":"hcloudmachinetemplate","controllerGroup":"infrastructure.cluster.x-k8s.io","controllerKind":"HCloudMachineTemplate","HCloudMachineTemplate":{"name":"my-cluster-controlplane","namespace":"default"},"namespace":"default","name":"my-cluster-controlplane","reconcileID":"f13c8e01-b69c-4731-ac9b-8dccc4fdc35a","HCloudMachineTemplate":{"name":"my-cluster-controlplane","namespace":"default"}}
{"level":"INFO","time":"2025-05-07T14:29:15.211Z","file":"controller/controller.go:110","message":"Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference","controller":"hetznercluster","controllerGroup":"infrastructure.cluster.x-k8s.io","controllerKind":"HetznerCluster","HetznerCluster":{"name":"my-cluster","namespace":"default"},"namespace":"default","name":"my-cluster","reconcileID":"85eab0d5-cbf7-4fb8-bae3-6f967bab82f7"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x144e678]

goroutine 385 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        sigs.k8s.io/controller-runtime@v0.18.7/pkg/internal/controller/controller.go:111 +0x19c
panic({0x16a3180?, 0x2bbdda0?})
        runtime/panic.go:791 +0x124
github.com/syself/cluster-api-provider-hetzner/pkg/services/hcloud/loadbalancer.createOptsFromSpec(0x400061d508)
        github.com/syself/cluster-api-provider-hetzner/pkg/services/hcloud/loadbalancer/loadbalancer.go:326 +0x1b8
github.com/syself/cluster-api-provider-hetzner/pkg/services/hcloud/loadbalancer.(*Service).createLoadBalancer(0x4000aa5828, {0x1c60e28, 0x40008152f0})
        github.com/syself/cluster-api-provider-hetzner/pkg/services/hcloud/loadbalancer/loadbalancer.go:290 +0x3c
github.com/syself/cluster-api-provider-hetzner/pkg/services/hcloud/loadbalancer.(*Service).Reconcile(0x4000aa5828, {0x1c60e28, 0x40008152f0})
        github.com/syself/cluster-api-provider-hetzner/pkg/services/hcloud/loadbalancer/loadbalancer.go:81 +0x1f0
github.com/syself/cluster-api-provider-hetzner/controllers.(*HetznerClusterReconciler).reconcileNormal(0x400038a380, {0x1c60e28, 0x40008152f0}, 0x400033a690)
        github.com/syself/cluster-api-provider-hetzner/controllers/hetznercluster_controller.go:198 +0x260
github.com/syself/cluster-api-provider-hetzner/controllers.(*HetznerClusterReconciler).Reconcile(0x400038a380, {0x1c60e28, 0x40008150b0}, {{{0x400000f4e0?, 0x5?}, {0x400000f4d0?, 0x400076dd08?}}})
        github.com/syself/cluster-api-provider-hetzner/controllers/hetznercluster_controller.go:173 +0x734
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1c656e8?, {0x1c60e28?, 0x40008150b0?}, {{{0x400000f4e0?, 0xb?}, {0x400000f4d0?, 0x0?}}})
        sigs.k8s.io/controller-runtime@v0.18.7/pkg/internal/controller/controller.go:114 +0x80
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0x4000622000, {0x1c60e60, 0x400024a2d0}, {0x177e7c0, 0x40001c7000})
        sigs.k8s.io/controller-runtime@v0.18.7/pkg/internal/controller/controller.go:311 +0x2d0
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x4000622000, {0x1c60e60, 0x400024a2d0})
        sigs.k8s.io/controller-runtime@v0.18.7/pkg/internal/controller/controller.go:261 +0x158
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        sigs.k8s.io/controller-runtime@v0.18.7/pkg/internal/controller/controller.go:222 +0x70
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 161
        sigs.k8s.io/controller-runtime@v0.18.7/pkg/internal/controller/controller.go:218 +0x3b8
```

--- a/pkg/services/hcloud/loadbalancer/loadbalancer.go
+++ b/pkg/services/hcloud/loadbalancer/loadbalancer.go
@@ -88,12 +88,12 @@ func (s *Service) Reconcile(ctx context.Context) (reconcile.Result, error) {
 			if errors.Is(err, ErrControlPlaneEndpointNotSet) {
 				// When an external ControlPlane Provider gets used (Kamaji), it might
 				// need some time until the endpoint is available.
-				err = fmt.Errorf("Requeue, waiting for control-plane endpoit to be set: %w",
+				err = fmt.Errorf("Requeue, waiting for control-plane endpoint to be set: %w",
 					err)
 				conditions.MarkFalse(
 					s.scope.HetznerCluster,
 					infrav1.LoadBalancerReadyCondition,
-					infrav1.LoadBalancerUpdateFailedReason,
+					infrav1.LoadBalancerUpdateFailedReason+"MissingControlPlaneEndpoint",
 					clusterv1.ConditionSeverityWarning,
 					"%s",
 					err.Error(),

--- a/pkg/services/hcloud/loadbalancer/loadbalancer.go
+++ b/pkg/services/hcloud/loadbalancer/loadbalancer.go
@@ -93,7 +93,7 @@ func (s *Service) Reconcile(ctx context.Context) (reconcile.Result, error) {
 				conditions.MarkFalse(
 					s.scope.HetznerCluster,
 					infrav1.LoadBalancerReadyCondition,
-					infrav1.LoadBalancerUpdateFailedReason+"MissingControlPlaneEndpoint",
+					"MissingControlPlaneEndpoint",
 					clusterv1.ConditionSeverityWarning,
 					"%s",
 					err.Error(),

--- a/pkg/services/hcloud/loadbalancer/loadbalancer.go
+++ b/pkg/services/hcloud/loadbalancer/loadbalancer.go
@@ -50,6 +50,9 @@ func NewService(scope *scope.ClusterScope) *Service {
 // ErrNoLoadBalancerAvailable indicates that no available load balancer could be fond.
 var ErrNoLoadBalancerAvailable = fmt.Errorf("no available load balancer")
 
+// ErrControlPlaneEndpointNotSet indicates that hetznercluster.spec.controlPlaneEndpoint is not set
+var ErrControlPlaneEndpointNotSet = errors.New("hetznercluster.spec.controlPlaneEndpoint is not set")
+
 // Reconcile implements the life cycle of HCloud load balancers.
 func (s *Service) Reconcile(ctx context.Context) (reconcile.Result, error) {
 	// delete the deprecated condition from existing cluster objects
@@ -77,11 +80,29 @@ func (s *Service) Reconcile(ctx context.Context) (reconcile.Result, error) {
 			if errors.Is(err, ErrNoLoadBalancerAvailable) {
 				return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
 			}
-		} else {
-			lb, err = s.createLoadBalancer(ctx)
+			return reconcile.Result{}, fmt.Errorf("failed to own existing load balancer (name=%s): %w", *s.scope.HetznerCluster.Spec.ControlPlaneLoadBalancer.Name, err)
 		}
+
+		lb, err = s.createLoadBalancer(ctx)
 		if err != nil {
-			return reconcile.Result{}, fmt.Errorf("failed to own/create load balancer: %w", err)
+			if errors.Is(err, ErrControlPlaneEndpointNotSet) {
+				// When an external ControlPlane Provider gets used (Kamaji), it might
+				// need some time until the endpoint is available.
+				err = fmt.Errorf("Requeue, waiting for control-plane endpoit to be set: %w",
+					err)
+				conditions.MarkFalse(
+					s.scope.HetznerCluster,
+					infrav1.LoadBalancerReadyCondition,
+					infrav1.LoadBalancerUpdateFailedReason,
+					clusterv1.ConditionSeverityWarning,
+					"%s",
+					err.Error(),
+				)
+				s.scope.Logger.Info(err.Error())
+				return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+
+			}
+			return reconcile.Result{}, fmt.Errorf("failed to create load balancer: %w", err)
 		}
 	}
 
@@ -327,7 +348,7 @@ func createOptsFromSpec(hc *infrav1.HetznerCluster) (hcloud.LoadBalancerCreateOp
 	}
 
 	if hc.Spec.ControlPlaneEndpoint == nil {
-		return hcloud.LoadBalancerCreateOpts{}, errors.New("hetznercluster.spec.controlPlaneEndpoint is not set")
+		return hcloud.LoadBalancerCreateOpts{}, ErrControlPlaneEndpointNotSet
 	}
 
 	listenPort := int(hc.Spec.ControlPlaneEndpoint.Port)

--- a/pkg/services/hcloud/loadbalancer/loadbalancer.go
+++ b/pkg/services/hcloud/loadbalancer/loadbalancer.go
@@ -50,7 +50,7 @@ func NewService(scope *scope.ClusterScope) *Service {
 // ErrNoLoadBalancerAvailable indicates that no available load balancer could be fond.
 var ErrNoLoadBalancerAvailable = fmt.Errorf("no available load balancer")
 
-// ErrControlPlaneEndpointNotSet indicates that hetznercluster.spec.controlPlaneEndpoint is not set
+// ErrControlPlaneEndpointNotSet indicates that hetznercluster.spec.controlPlaneEndpoint is not set.
 var ErrControlPlaneEndpointNotSet = errors.New("hetznercluster.spec.controlPlaneEndpoint is not set")
 
 // Reconcile implements the life cycle of HCloud load balancers.
@@ -71,38 +71,37 @@ func (s *Service) Reconcile(ctx context.Context) (reconcile.Result, error) {
 	}
 
 	if lb == nil {
-		// fixed name is set - we expect a load balancer with this name to exist
-
 		if s.scope.HetznerCluster.Spec.ControlPlaneLoadBalancer.Name != nil {
+			// fixed name is set - we expect a load balancer with this name to exist
 			lb, err = s.ownExistingLoadBalancer(ctx)
-
-			// if load balancer is not found even though we expect it to exist, wait and reconcile until user creates it
-			if errors.Is(err, ErrNoLoadBalancerAvailable) {
-				return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
+			if err != nil {
+				// if load balancer is not found even though we expect it to exist, wait and reconcile until user creates it
+				if errors.Is(err, ErrNoLoadBalancerAvailable) {
+					return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
+				}
+				return reconcile.Result{}, fmt.Errorf("failed to own existing load balancer (name=%s): %w", *s.scope.HetznerCluster.Spec.ControlPlaneLoadBalancer.Name, err)
 			}
-			return reconcile.Result{}, fmt.Errorf("failed to own existing load balancer (name=%s): %w", *s.scope.HetznerCluster.Spec.ControlPlaneLoadBalancer.Name, err)
-		}
-
-		lb, err = s.createLoadBalancer(ctx)
-		if err != nil {
-			if errors.Is(err, ErrControlPlaneEndpointNotSet) {
-				// When an external ControlPlane Provider gets used (Kamaji), it might
-				// need some time until the endpoint is available.
-				err = fmt.Errorf("Requeue, waiting for control-plane endpoint to be set: %w",
-					err)
-				conditions.MarkFalse(
-					s.scope.HetznerCluster,
-					infrav1.LoadBalancerReadyCondition,
-					"MissingControlPlaneEndpoint",
-					clusterv1.ConditionSeverityWarning,
-					"%s",
-					err.Error(),
-				)
-				s.scope.Logger.Info(err.Error())
-				return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
-
+		} else {
+			lb, err = s.createLoadBalancer(ctx)
+			if err != nil {
+				if errors.Is(err, ErrControlPlaneEndpointNotSet) {
+					// When an external ControlPlane Provider gets used (Kamaji), it might
+					// need some time until the endpoint is available.
+					err = fmt.Errorf("requeue, waiting for control-plane endpoint to be set: %w",
+						err)
+					conditions.MarkFalse(
+						s.scope.HetznerCluster,
+						infrav1.LoadBalancerReadyCondition,
+						"MissingControlPlaneEndpoint",
+						clusterv1.ConditionSeverityWarning,
+						"%s",
+						err.Error(),
+					)
+					s.scope.Logger.Info(err.Error())
+					return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+				}
+				return reconcile.Result{}, fmt.Errorf("failed to create load balancer: %w", err)
 			}
-			return reconcile.Result{}, fmt.Errorf("failed to create load balancer: %w", err)
 		}
 	}
 

--- a/pkg/services/hcloud/loadbalancer/loadbalancer.go
+++ b/pkg/services/hcloud/loadbalancer/loadbalancer.go
@@ -287,7 +287,10 @@ func (s *Service) reconcileServices(ctx context.Context, lb *hcloud.LoadBalancer
 }
 
 func (s *Service) createLoadBalancer(ctx context.Context) (*hcloud.LoadBalancer, error) {
-	opts := createOptsFromSpec(s.scope.HetznerCluster)
+	opts, err := createOptsFromSpec(s.scope.HetznerCluster)
+	if err != nil {
+		return nil, err
+	}
 	lb, err := s.scope.HCloudClient.CreateLoadBalancer(ctx, opts)
 	if err != nil {
 		err = fmt.Errorf("failed to create load balancer: %w", err)
@@ -309,7 +312,7 @@ func (s *Service) createLoadBalancer(ctx context.Context) (*hcloud.LoadBalancer,
 	return lb, nil
 }
 
-func createOptsFromSpec(hc *infrav1.HetznerCluster) hcloud.LoadBalancerCreateOpts {
+func createOptsFromSpec(hc *infrav1.HetznerCluster) (hcloud.LoadBalancerCreateOpts, error) {
 	// gather algorithm type
 	algorithmType := hc.Spec.ControlPlaneLoadBalancer.Algorithm.HCloudAlgorithmType()
 
@@ -321,6 +324,10 @@ func createOptsFromSpec(hc *infrav1.HetznerCluster) hcloud.LoadBalancerCreateOpt
 	var network *hcloud.Network
 	if hc.Status.Network != nil {
 		network = &hcloud.Network{ID: hc.Status.Network.ID}
+	}
+
+	if hc.Spec.ControlPlaneEndpoint == nil {
+		return hcloud.LoadBalancerCreateOpts{}, errors.New("hetznercluster.spec.controlPlaneEndpoint is not set")
 	}
 
 	listenPort := int(hc.Spec.ControlPlaneEndpoint.Port)
@@ -341,7 +348,7 @@ func createOptsFromSpec(hc *infrav1.HetznerCluster) hcloud.LoadBalancerCreateOpt
 				Proxyprotocol:   &proxyprotocol,
 			},
 		},
-	}
+	}, nil
 }
 
 // Delete implements the deletion of HCloud load balancers.

--- a/pkg/services/hcloud/loadbalancer/loadbalancer_test.go
+++ b/pkg/services/hcloud/loadbalancer/loadbalancer_test.go
@@ -122,7 +122,8 @@ var _ = Describe("createOptsFromSpec", func() {
 		hetznerCluster.Status.Network = nil
 		wantCreateOpts.Network = nil
 
-		createOpts := createOptsFromSpec(hetznerCluster)
+		createOpts, err := createOptsFromSpec(hetznerCluster)
+		Expect(err).To(BeNil())
 
 		// ignore random name
 		createOpts.Name = ""
@@ -131,7 +132,8 @@ var _ = Describe("createOptsFromSpec", func() {
 	})
 
 	It("creates specs for cluster with network", func() {
-		createOpts := createOptsFromSpec(hetznerCluster)
+		createOpts, err := createOptsFromSpec(hetznerCluster)
+		Expect(err).To(BeNil())
 
 		// ignore random name
 		createOpts.Name = ""
@@ -142,7 +144,8 @@ var _ = Describe("createOptsFromSpec", func() {
 	It("creates specs for cluster without load balancer name set", func() {
 		hetznerCluster.Spec.ControlPlaneLoadBalancer.Name = nil
 
-		createOpts := createOptsFromSpec(hetznerCluster)
+		createOpts, err := createOptsFromSpec(hetznerCluster)
+		Expect(err).To(BeNil())
 
 		// should generate correct name
 		Expect(createOpts.Name).To(HavePrefix("hetzner-cluster-kube-apiserver-"))

--- a/pkg/services/hcloud/loadbalancer/loadbalancer_test.go
+++ b/pkg/services/hcloud/loadbalancer/loadbalancer_test.go
@@ -17,6 +17,8 @@ limitations under the License.
 package loadbalancer
 
 import (
+	"errors"
+
 	"github.com/go-logr/logr"
 	"github.com/hetznercloud/hcloud-go/v2/hcloud"
 	. "github.com/onsi/ginkgo/v2"
@@ -154,5 +156,12 @@ var _ = Describe("createOptsFromSpec", func() {
 		createOpts.Name = ""
 		wantCreateOpts.Name = ""
 		Expect(createOpts).To(Equal(wantCreateOpts))
+	})
+
+	It("returns ErrControlPlaneEndpointNotSet", func() {
+		hetznerCluster.Spec.ControlPlaneEndpoint = nil
+
+		_, err := createOptsFromSpec(hetznerCluster)
+		Expect(errors.Is(err, ErrControlPlaneEndpointNotSet)).To(BeTrue())
 	})
 })

